Added optional support of regex in XPath

devova opened the pull request:
I as a developer would like to specify regex in XPath queries for more precise search.
Example: `//targets/Name[not(re:match(@id,'pipeline$'))`

hchasestevens commented on astpath/search.py:
```diff
@@ -21,13 +21,13 @@ class XMLVersions:
 
 
-def _query_factory(verbose=False):
+def _query_factory(verbose=False, allow_re=False):
```
I think we would like to enable this by default.

devova commented:
Hello. I just addressed your comments. I'm gonna create followup PR with test, when your become ready. Regarding bumping versions, I recommend to use `setuptools_scm` that will determinate version based on git tag.

hchasestevens commented:
Hi @devova ,

Sorry for the slow progress on this. I am having difficulty verifying your feature, can you please indicate where I am making a mistake? I believe your change will add a lot of utility to the tool and so would like to get it working out-of-the-box for all users.

```bash
$ docker run --rm -it python:3.6 bash
root@038e3d92902a:/# pip install lxml
Collecting lxml
  Downloading https://files.pythonhosted.org/packages/03/a4/9eea8035fc7c7670e5eab97f34ff2ef0ddd78a491bf96df5accedb0e63f5/lxml-4.2.5-cp36-cp36m-manylinux1_x86_64.whl (5.8MB)
    100% |████████████████████████████████| 5.8MB 6.4MB/s 
Installing collected packages: lxml
Successfully installed lxml-4.2.5
root@038e3d92902a:/# pip install git+https://github.com/hchasestevens/astpath.git@refs/pull/8/head
Collecting git+https://github.com/hchasestevens/astpath.git@refs/pull/8/head
  Cloning https://github.com/hchasestevens/astpath.git (to revision refs/pull/8/head) to /tmp/pip-req-build-mmme6ke3
  Did not find branch or tag 'refs/pull/8/head', assuming revision or ref.
Building wheels for collected packages: astpath
  Running setup.py bdist_wheel for astpath ... done
  Stored in directory: /tmp/pip-ephem-wheel-cache-hr3xeth9/wheels/16/d5/85/3961b7efa7f9f9e94fa46e5284bd9e822cc4cf13e4e326bc71
Successfully built astpath
Installing collected packages: astpath
Successfully installed astpath-0.9.0
root@038e3d92902a:/# cd /usr/local/lib/python3.6/
root@038e3d92902a:/usr/local/lib/python3.6# astpath "//targets/Name[not(re:match(@id,'pipeline$'))"
Traceback (most recent call last):
  File "/usr/local/bin/astpath", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/astpath/cli.py", line 59, in main
    allow_ns=args.regular_expressions
  File "/usr/local/lib/python3.6/site-packages/astpath/search.py", line 157, in search
    matching_elements = query(xml_ast, expression)
  File "/usr/local/lib/python3.6/site-packages/astpath/search.py", line 31, in lxml_query
    return element.xpath(expression, namespaces=namespaces)
  File "src/lxml/etree.pyx", line 1577, in lxml.etree._Element.xpath
  File "src/lxml/xpath.pxi", line 307, in lxml.etree.XPathElementEvaluator.__call__
  File "src/lxml/xpath.pxi", line 227, in lxml.etree._XPathEvaluatorBase._handle_result
lxml.etree.XPathEvalError: Invalid predicate
```

Thanks,
Chase

hchasestevens commented:
Hi @devova , I'm closing this as the requested feature is now available in version 0.9.0. The syntax is as follows:
```bash
astpath "//Name[re:match('*.foo$', @id)]"
astpath "//Name[re:search('bar\d+', @id)]"
```
Hope this helps!
